File: Python/programmers/etc/level3/등대.py
import sys
sys.setrecursionlimit(10**6)

def solution(n, lighthouse):
    
    def dfs(node):
        visited[node] = 1
        if not graph[node]:
            return 1, 0
        on, off = 1, 0
        for nxt in graph[node]:
            if not visited[nxt]:
                child_on, child_off = dfs(nxt)
                on += min(child_on, child_off)
                off += child_on
        return on, off
    
    graph = [[] for _ in range(n+1)]
    for u, v in lighthouse:
        graph[u].append(v)
        graph[v].append(u)
    visited = [0]*(n+1)
    
    on, off = dfs(1)
    
    return min(on, off)


n = 8
lighthouse = [[1, 2], [1, 3], [1, 4], [1, 5], [5, 6], [5, 7], [5, 8]]
print(solution(n, lighthouse)) # 2

n = 10
lighthouse = [[4, 1], [5, 1], [5, 6], [7, 6], [1, 2], [1, 3], [6, 8], [2, 9], [9, 10]]
print(solution(n, lighthouse)) # 3

File: Java/programmers/etc/level3/등대.java
package CodeTest.Java.programmers.etc.level3;

import java.util.*;

public class 등대 {
    private static List<Set<Integer>> arr;
    private static int[] visited;

    public int solution(int n, int[][] lighthouse) {
        int answer = 0;
        arr = new ArrayList<>();
        
        for (int i=0; i<n; i++) {
            arr.add(new HashSet<>());
        }

        for (int[] node : lighthouse) {
            arr.get(node[0]-1).add(node[1]-1);
            arr.get(node[1]-1).add(node[0]-1);
        }

        visited = new int[n];
        
        dfs(0);
        
        for (int v : visited) {
            if (v == 2) answer++;
        }

        return answer;
    }

    private static void dfs(int cur) {
        visited[cur] = 1;
        Set<Integer> set = arr.get(cur);
        for (int nxt : set) {
            if (visited[nxt] == 0) {
                dfs(nxt);
                if (visited[nxt] == 1) {
                    visited[nxt] = 2;
                }
            }
        }
    }
}
